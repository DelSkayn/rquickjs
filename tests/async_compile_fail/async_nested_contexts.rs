use rquickjs::{async_with, AsyncContext, AsyncRuntime};

pub async fn test() {
    let rt = AsyncRuntime::new().unwrap();
    let ctx_1 = AsyncContext::full(&rt).await.unwrap();
    let ctx_2 = AsyncContext::full(&rt).await.unwrap();
    async_with!(ctx_1 => |ctx_1|{
        async_with!(ctx_2 => |ctx_2|{
            // It is disallowed to use multiple ctx object together from different with closures.
            // Lifetime on ctx should be unique.
            ctx_1.globals().set("t", ctx_2.globals());
        }).await
    })
    .await
}

fn main() {}
