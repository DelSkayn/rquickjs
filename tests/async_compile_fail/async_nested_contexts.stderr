error[E0521]: borrowed data escapes outside of closure
  --> tests/async_compile_fail/async_nested_contexts.rs:11:38
   |
7  | /      async_with!(ctx_1 => |ctx_1|{
8  | |/         async_with!(ctx_2 => |ctx_2|{
9  | ||             // It is disallowed to use multiple ctx object together from different with closures.
10 | ||             // Lifetime on ctx should be unique.
11 | ||             ctx_1.globals().set("t", ctx_2.globals());
   | ||                                      ^^^^^^^^^^^^^^^ `ctx_2` escapes the closure body here
12 | ||         }).await
   | ||__________- `ctx_2` is a reference that is only valid in the closure body
13 | |      })
   | |_______- `ctx_1` declared here, outside of the closure body

error[E0521]: borrowed data escapes outside of closure
  --> tests/async_compile_fail/async_nested_contexts.rs:11:13
   |
7  | /     async_with!(ctx_1 => |ctx_1|{
8  | |         async_with!(ctx_2 => |ctx_2|{
9  | |             // It is disallowed to use multiple ctx object together from different with closures.
10 | |             // Lifetime on ctx should be unique.
11 | |             ctx_1.globals().set("t", ctx_2.globals());
   | |             ^^^^^^^^^^^^^^^
   | |             |
   | |             `ctx_1` escapes the closure body here
   | |             argument requires that `'1` must outlive `'static`
12 | |         }).await
13 | |     })
   | |      -
   | |      |
   | |______`ctx_1` is a reference that is only valid in the closure body
   |        has type `Ctx<'1>`
